doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,283 @@
+* backlink to the owning user (that created it) - this backlink must be
+  maintained; inspiration can be seen in working with hostgroups
+
+The definition of the new object and its attributes holding application-specific
+passwords and related information, as well as operations with them (add, modify,
+delete) will be contained in a new file - `ipaserver/plugins/appspecificpw.py`.
+
+Names of the newly created objects in the new subtree will somehow have to be
+validated and enforced (possibly, but not necessarily using an LDAP plugin), so
+that the user will not be able to create a principal looking like a host or a
+service, or even set a password for another user.
+
+The existing mechanism for generating passwords will be made available for use
+for application-specific passwords as well. Of course, the user will be able to
+set application-specific passwords as usual (having to fulfil password policies
+in place).
+
+Password reset functionality must be extended to facilitate multiple passwords
+for a single user as well - the user will be able to specify what password to
+reset using the application name of the application-specific password. If it
+will be left blank, then primary password will be reset, otherwise
+application-specific password for the respective application will be reset. A
+new parameter therefore needs to be added to the password reset command -
+application name (for the application-specific password).
+
+The user account status will display information for all its passwords.
+
+### ipa-kdb support for the aforementioned functionality
+
+The ipa-kdb daemon will be extended to support application-specific passwords
+as well, so that they can be used with Kerberos and HBAC.
+
+Subaccounts will be introduced with principals in the form 
+```_appname_/_user_@_REALM_```
+(to be consistent with other things in FreeIPA) - a username will remain the
+same for subaccounts, only the principal will contain the application name. The
+code of FreeIPA must be modified to take into account the newly changed
+assumptions. Until now, these were:
+* two-component principals with a host as the first component are always host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are always service principals
+  (`_service_/_fqdn_`)
+* single-component principals are always users (user)
+Now, they will be:
+* two-component principals with an application name as the first component are
+  users logged in with an application-specific password (`_appname_/_user_`)
+* two-component principals with a host as the first component are host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are service principals (`_service_/_fqdn_`)
+* single-component principals are users logged in with their primary password
+  (`user`)
+
+### Permissions management for application-specific passwords
+
+New ACIs must automatically be added and maintained for every
```

Comment:
Clearly ACIs need to be  built so that a static set will handle this properly.
ACIs are expensive, adding them dynamically would kill performance very quickly and become unmanageable, esp if any administrative change is needed.
It is easy to create a set of ACIs that will just work and do what is recommended here w/o dynamic creation.

Reply:
All right, I will create a static set of ACIs. It may even be simpler to implement.